common/persistence/visibility/store/elasticsearch/processor.go

```diff
@@ -172,7 +179,18 @@ func (p *processorImpl) hashFn(key interface{}) uint32 {
 
 // Add request to the bulk and return a future object which will receive ack signal when request is processed.
 func (p *processorImpl) Add(request *client.BulkableRequest, visibilityTaskKey string) *future.FutureImpl[bool] {
-	newFuture := newAckFuture()
+	newFuture := newAckFuture() // Create future first to measure impact of following RWLock on latency
+
+	p.shutdownLock.RLock()
+	defer p.shutdownLock.RUnlock()
+
+	if atomic.LoadInt32(&p.status) == common.DaemonStatusStopped {
+		p.logger.Warn("Rejecting ES request for visibility task key because processor has been shut down.", tag.Key(visibilityTaskKey), tag.ESDocID(request.ID), tag.Value(request.Doc))
+		errFuture := future.NewFuture[bool]()
+		errFuture.Set(false, errVisibilityShutdown)
+		return errFuture
```

Comment:
nit: now we can reuse the future in `newFuture` and avoid allocating a new channel.

Not related to your PR but it seems like the assignment on L202 will cause the underlying channel in `newFuture` not getting closed. I am not sure if golang GC can handle this case or result in some leak. cc @alexshtin